server/config.go

```diff
@@ -0,0 +1,185 @@
+		return nil, err
+	}
+
+	return placeholder, nil
+}
+
+// ConfigurationStorage holds one proposer configuration.
+type ConfigurationStorage struct {
+	FeeRecipient types.Address
+	Enabled      bool
+	Relays       []RelayEntry
+	GasLimit     types.U256Str
+}
+
+// ProposerConfigurationStorage holds both the default configuration and the proposers ones.
+type ProposerConfigurationStorage struct {
+	proposerConfigurations map[types.PublicKey]*ConfigurationStorage
+	defaultConfiguration   *ConfigurationStorage
+}
+
+// NewProposerConfigurationStorage creates a new storage holding each proposer preferences using
+// the content extracted from a JSON file.
+func NewProposerConfigurationStorage(filename string) (*ProposerConfigurationStorage, error) {
+	// Tries to create the raw configuration extracted from the JSON file.
+	raw, err := newRawConfigurationFile(filename)
+	if err != nil {
+		return nil, err
+	}
+
+	// Initialize the storage and save default configuration.
+	pcs := &ProposerConfigurationStorage{
+		proposerConfigurations: map[types.PublicKey]*ConfigurationStorage{},
+	}
+
+	pcs.defaultConfiguration, err = newConfigurationStorage(&raw.DefaultConfig, raw.BuilderRelaysGroups)
+	if err != nil {
+		return nil, err
+	}
+
+	// For each proposer, save its own configuration.
+	for proposer, configuration := range raw.ProposerConfig {
+		address, err := types.HexToPubkey(proposer)
+		if err != nil {
+			return nil, err
+		}
+
+		configurationStorage, err := newConfigurationStorage(&configuration, raw.BuilderRelaysGroups)
+		if err != nil {
+			return nil, err
+		}
+
+		pcs.proposerConfigurations[address] = configurationStorage
+	}
+
+	return pcs, nil
+}
+
+// GetProposerConfiguration looks for a specific configuration for the given proposer, if not found it
+// returns the default configuration.
+func (s *ProposerConfigurationStorage) GetProposerConfiguration(proposer types.PublicKey) *ConfigurationStorage {
+	res := s.proposerConfigurations[proposer]
+	if res == nil {
+		res = s.defaultConfiguration
+	}
+
+	return res
+}
+
+// newConfigurationStorage creates a new ConfigurationStorage from a rawConfiguration
+// previously extracted from a JSON file and the relay groups available.
+// Used to create the default configuration and each proposer's one.
+func newConfigurationStorage(rawConf *rawConfiguration, groups map[string][]string) (*ConfigurationStorage, error) {
+	feeRecipient, err := types.HexToAddress(rawConf.FeeRecipient)
+	if err != nil {
+		return nil, err
+	}
+
+	gasLimit := types.U256Str{}
+	err = gasLimit.UnmarshalText([]byte(rawConf.ValidatorRegistration.GasLimit))
+	if err != nil {
+		return nil, err
+	}
+
+	configuration := &ConfigurationStorage{
+		FeeRecipient: feeRecipient,
+		Enabled:      rawConf.ValidatorRegistration.Enabled,
+		GasLimit:     gasLimit,
+	}
+
+	if len(rawConf.ValidatorRegistration.BuilderRelays) == 0 {
+		return nil, errors.New("no builder is associated to this proposer")
```

Comment:
this might actually be allowed, ping @tbenr

Reply:
If it refers to a specific pubkey, yes I think it should be optional. 
The idea is that default config must have the relay set defined and specific pubkey config can optionally override it. 
cc @StefanBratanov

Reply:
cc @0xpanoramix -- shouldn't this just fall back to the default config here rather than return an error?

Reply:
It definitely should, good catch thanks !